bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/JRubyEngineWrapper.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.automation.jrubyscripting.internal;
+
+import java.io.Reader;
+
+import javax.script.Bindings;
+import javax.script.Compilable;
+import javax.script.CompiledScript;
+import javax.script.Invocable;
+import javax.script.ScriptContext;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineFactory;
+import javax.script.ScriptException;
+
+import org.jruby.embed.jsr223.JRubyEngine;
+
+/**
+ * This is a wrapper for {@link JRubyEngine}.
+ * 
+ * The purpose of this class is to intercept the call to eval and save the context into
+ * a global variable for use in the helper library.
+ *
+ * @author Jimmy Tanagra - Initial contribution
+ */
+public class JRubyEngineWrapper implements Compilable, Invocable, ScriptEngine {
+
+    private final JRubyEngine engine;
+
+    private final String CONTEXT_VAR_NAME = "$ctx";
+
+    JRubyEngineWrapper(JRubyEngine engine) {
+        this.engine = engine;
+    }
+
+    @Override
+    public CompiledScript compile(String script) throws ScriptException {
+        return engine.compile(script);
+    }
+
+    @Override
+    public CompiledScript compile(Reader reader) throws ScriptException {
+        return engine.compile(reader);
+    }
+
+    @Override
+    public Object eval(String script, ScriptContext context) throws ScriptException {
+        Object ctx = context.getBindings(ScriptContext.ENGINE_SCOPE).get("ctx");
```

Comment:
eval is used in various places including inside the addon itself to inject the require, setting ruby environment var, etc. We don't want to make all those local variables global.

We could get around that by adding a `doEval` or `internalEval` that just directly delegates it to the real engine eval, bypassing the globalizer. Then call this `doEval` within the addon.

This leaves just the core calling eval which includes the transformations / profile too. If we go through all the variables in the binding, it would make e.g. variables in transformation scripts also global. Is that OK?

Reply:
Okay, we can leave as a single special case. Though I would encourage having a constant for "ctx", and _not_ for "$ctx", and prepending a $ to the former when setting it below. That will make it easier if/when we need to make it a list in the future.